templates/node/node.py

```diff
@@ -0,0 +1,277 @@
+
+    def validate(self):
+        if not self.data['vdc']:
+            raise RuntimeError('vdc name should be given')
+        # ensure uploaded key
+        self.sshkey
+
+    def update_data(self, data):
+        # merge the new data
+        self.data.update(data)
+        self.save()
+
+    @property
+    def ovc(self):
+        """
+        An ovc connection instance
+        """
+        if self._ovc is not None:
+            return self._ovc
+        
+        vdc = self.vdc
+        instance = vdc.ovc.instance
+        self._ovc = j.clients.openvcloud.get(instance=instance)
+
+        return self._ovc
+
+    @property
+    def vdc(self):
+        if self._vdc:
+            return self._vdc
+
+        # Get object for an VDC service, make sure exactly one is running
+        vdc = self.api.services.find(template_uid=self.VDC_TEMPLATE, name=self.data['vdc'])
+        if len(vdc) != 1:
+            raise RuntimeError('found %s vdc, requires exactly 1' % len(vdc))
+        
+        self._vdc = vdc[0]
+
+        return self._vdc
+
+    @property
+    def space(self):
+        return self.ovc.space_get(
+            accountName=None,
+            spaceName=self.data['vdc']
+            )
+
+    @property
+    def sshkey(self):
+        """ Get a path and keyname of the sshkey service """
+
+        sshkeys = self.api.services.find(template_uid=self.SSH_TEMPLATE)
+        if len(sshkeys) == 0:
+            raise RuntimeError('no %s ssh services found' % len(sshkeys))
+
+        # Get key name and path
+        path = sshkeys[0].data['path']
+        key = path.split('/')[-1]
+
+        return key         
+
+    @property
+    def machine(self):
+        if self._machine:
+            return self._machine
+        return self.space.machines.get(self.name)
+
+    def install(self):
+        try:
+            self.state.check('actions', 'install', 'ok')
+            return
+        except StateCheckError:
+            pass        
+
+        machine = self.machine
+        if not machine:
+            machine = self._machine_create()
+
+        # Get data from the vm
+        ip_private, vm_info = machine.machineip_get()
+        self.data['sshLogin'] = vm_info['accounts'][0]['login']
+        self.data['sshPassword']= vm_info['accounts'][0]['password']
+        self.data['ipPrivate'] = ip_private
+        self.data['ipPublic'] = machine.space.model['publicipaddress']
+        self.data['machineId'] = machine.id
+
+        self.portforward_create(self.data['ports'])
+        self._configure_disks()
+        self.save()
+
+        self.state.set('actions', 'install', 'ok')
+
+    def _machine_create(self):
+        """ Create a new machine """
+        data = self.data
+        space = self.space
+        self._machine = space.machine_create(
+            name=self.name,
+            sshkeyname= self.sshkey,
+            image=data['osImage'],
+            disksize=data['bootDiskSize'],
+            datadisks=[data['dataDiskSize']],
```

Comment:
is `dataDiskSize` mandatory ? I think it should be optional

Reply:
Default for `dataDiskSize=10 GB` is given in capnp schema